Stacks update to 2.4

bernt-matthias opened the pull request:
Upgrade of the stacks suite to latest 2.4

TODOs: 

- [x] fix WRAPPER_VERSION when https://github.com/bioconda/bioconda-recipes/pull/12420 is merged
- [x] move to stacks2 directory (I would suggest to have parallel stacks and stacks2 tools available)
- [x] clean test data
- [x] make all tests work
- [x] reword black/whitelist as in https://github.com/galaxyproject/tools-iuc/pull/2124
- [x] change owner in shed.yml to iuc 
- [x] check help texts
- [x] remove todo.txt

Topics for discussion (might be out of date at the moment):

clone-/kmerfilter
- [x] add proper pair as input option
- [x] log file output
- [x] kmer: allow kmer freq table as input
  - wont do this (now)
- [x] kmer: add kmer freq distribution output

procrad:
- [x] remove output type "auto": we had this discussion for samtools and decided against it, because its difficult/impossible to use in workflows
  - Maybe we leave it as is. Currently, irrespective of the chosen output type the data sets and their type is autodetected. So it should not make a difference. Needs to be changes only if this turns out unusable in workflows...
- [x] according to help -b is optional ()
  - it isn't .. I tested it
- [x] allow FASTA, BAM input (latter probably via -f?)
  - leave for future releases
- [x] with -E also other fastq types would be possible
  - leave for future releases
- [x] --merge useful (at least not with current tool)?
  - leave for future releases
- [x] add more structure to the advanced section? The structure of `process_radtags -h`could be adapted (unfortunately short options are not structured). 
  - leave for future releases

cstacks:
- [x] support `--catalog`?
  - leave for future releases

- [x] upgrade from 2.1 to 2.2 https://github.com/bioconda/bioconda-recipes/pull/10942
- [x] procrad: tests for the new options?

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

FOR REVIEWER:
* [ ] .shed.yml file ok
    - [ ] Toolshed user `iuc` has access to associated toolshed repo(s)
* [ ] Indentation is correct (4 spaces)
* [ ] Tool version/build ok
* [ ] `<command/>`
  - [ ] Text parameters, input and output files `'single quoted'`
  - [ ] Use of `<![CDATA[ ... ]]>` tags
  - [ ] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [ ] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [ ] Tests
  - [ ] Parameters are reasonably covered
  - [ ] Test files are appropriate
* [ ] Help
  - [ ] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [ ] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

bernt-matthias commented:
Hey @abretaud and @yvanlebras. Made some progress here: I'm optimistic that all tools except for  clone_filter, kmer_filter und process_shortreads work now (i.e. the extensive test cases for all other tools work). Would be happy if you could give me some comments. 

I have uploaded the tools to the testtoolshed, but the perl scripts are dysfunctional in the available bioconda builds (https://github.com/bioconda/bioconda-recipes/pull/12420) in order to get them running you just need to quote the '@' in the exe_path variable.

bernt-matthias commented:
For some reason suite_stacks2 is unfortunately not created in the testtoolshed

abretaud commented:
Thanks for your work on this @bernt-matthias! I'll review soon
Could you resolve the conflict on tools/stacks/stacks_genotypes.xml?
I wonder if we should keep the stacks dir and create a new stacks2 dir to keep track of the old xmls. I guess there won't be a new stacks 1.xx release, but if we want to fix some bugs in the 1.xx tools, it could be useful to keep them.

bernt-matthias commented:
> I wonder if we should keep the stacks dir and create a new stacks2 dir to keep track of the old xmls.

Definitely. It's on the very top of my TODO list. This will also resolve the conflict (which is because genotypes is absent in the new version).

yvanlebras commented:
Thank you @bernt-matthias @abretaud to work on it! I can test it when TS repos directly usable if of interest!

bernt-matthias commented:
Everything is in the testtoolshed and should be usable.

bernt-matthias commented:
did some fixes (one critical). new versions are in the testtoolshed.

bernt-matthias commented:
Dear @yvanlebras @abretaud just one question: I'm looking for reasonable names for the output collections of the pipeline components. I would like to have something like TOOL on DATA: DESCRIPTION. TOOL on DATA is easy, so I'm looking for the DESCRIPTION part.  Below you find the contents of the collections and my suggestions for the descriptions. 

(One more comment. I adapted the strategy of the old stacks wrappers to store all outputs from a tool in a collection. This could be changed to nested collections, which would make it easier to find reasonable descriptions. What do you think?) 

**ustacks** 

- the collections contains the SAMPLE.tags.tsv, SAMPLE.alleles.tsv, and SAMPLE.snps.tsv
- According to the docs this is Assembled loci, Model calls from each locus, Haplotypes/alleles recorded from each locus
- Suggestion: _Loci per sample_

**cstacks**

- contains catalog.tags.tsv, catalog.snps.tsv, catalog.alleles.tsv
- suggestion: _Catalog of loci_ 

**sstacks**

- contains SAMPLE.matches.tsv
- suggestion _Matches to the catalog_ 

**tsv2bam**
- contains SAMPLE.matches).bam
- suggestion: _Loci_

**gstacks**
- contains catalog.calls.vcf and catalog.fa.gz
- suggestion: _Assembled contigs and variant sites_

**populations**
- I took the descriptions from the manual. Should be fine.

yvanlebras commented:
Thank you for working on it! 
Considering previous questions, it seems to me that something like:
- "Loci and polymorphism" is maybe better for ustack
- "Matches to the catalog (bam)" is better for description of tsv2bam step. 

Using nested collections can be better, but a major (old?) issue using collections is when there is more than 10 000 files generated on it .. Do you think nested collections can additionally help on this ?

Not ready to test new versions for now but don't hesitate to ping me when you think this can be of interest..

yvanlebras commented:
... and amazing happy new year Bernt!

bgruening commented:
@bernt-matthias is this still WIP and can we bribe @abretaud and @yvanlebras to review it? :)

bernt-matthias commented:
Ready for review from my side.

bernt-matthias commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,439 @@
+    <token name="@STACKS_VERSION@">2.2</token>
+    <token name="@WRAPPER_VERSION@">2</token>
+
+    <xml name="stdio">
```
yep. done

bernt-matthias commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,439 @@
+    </xml>
+
+    <token name="@STACKS_VERSION@">2.2</token>
+    <token name="@WRAPPER_VERSION@">2</token>
```
yep .. had this just for the test TS. done

bernt-matthias commented on tools/stacks2/news.txt:
```diff
@@ -0,0 +1,181 @@
+Stacks 2.2 - Aug 22, 2018
+--------------------------
```
sure

bernt-matthias commented on tools/stacks2/stacks_shortreads.xml:
```diff
@@ -0,0 +1,234 @@
+                </element>
+            </output_collection>
+        </test>
+<!--        <test>-->
```
gone. done

bernt-matthias commented on tools/stacks2/stacks_kmerfilter.xml:
```diff
@@ -0,0 +1,180 @@
+            </section>
+            <output name="kfreq" file="kmerfilter/kfreq.tsv"/>
+        </test>
+	<!--<test>
```
gone. done.

bernt-matthias commented:
Did some improvements in the filter tools (some of which break it at the moment). Bugs have been reported upstream https://groups.google.com/forum/#!topic/stacks-users/GZi_Cehs9_U ... 

So for the moment everything but the filters seem ready to review.

bernt-matthias commented:
stacks 2.3 is out https://groups.google.com/forum/#!msg/stacks-users/QYW8Pd15FI8/E0-Dyh0bAwAJ

since only gstacks and populations seem affected I will give it a try ..

abretaud commented on tools/stacks2/macros_process.xml:
```diff
@@ -0,0 +1,235 @@
+                <option value="-r">yes</option>
+                <option value="" selected="true">no</option>
+            </param>
+	    <when value="-r">
```
tabs here and in the following line

bernt-matthias replied:
done

abretaud commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,454 @@
+        <option value="apeKI">apeKI</option>
+        <option value="apoI">apoI</option>
+        <option value="aseI">aseI</option>
+	<option value="bamHI">bamHI</option>
```
tabs (and below)

bernt-matthias replied:
done

abretaud commented on tools/stacks2/stacks_cstacks.xml:
```diff
@@ -0,0 +1,175 @@
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/>
```
tabs

bernt-matthias replied:
done

abretaud commented on tools/stacks2/stacks_gstacks.xml:
```diff
@@ -0,0 +1,340 @@
+        </test>
+        <!-- refbased w popmap (here bam names need to be equal to sample names in popmap), \-\-unpaired, advanced, snp model -->
+        <test>
+		<param name="input_bam" ftype="bam" value="refmap/PopA_01.bam,refmap/PopA_02.bam"/>
```
tabs

bernt-matthias replied:
done

abretaud commented on tools/stacks2/stacks_kmerfilter.xml:
```diff
@@ -0,0 +1,217 @@
+    ]]></command>
+    <inputs>
+        <expand macro="fastq_input_pair_bc"/>
+	<param name="capture" type="boolean" checked="false" truevalue="-D" falsevalue="" argument="-D" label="Capture discarded reads to a file" />
```
tabs

bernt-matthias replied:
done

abretaud commented on tools/stacks2/stacks_procrad.xml:
```diff
@@ -0,0 +1,288 @@
+
+        <section name="options_advanced" title="advanced options" expanded="False">
+	    <expand macro="common_advanced"/>
+            <param argument="--bestrad" type="boolean" checked="false" truevalue="--bestrad" falsevalue="" label="library was generated using BestRAD, check for restriction enzyme on either read and potentially tranpose reads" />
```
tabs

bernt-matthias replied:
done

abretaud commented on tools/stacks2/stacks_shortreads.xml:
```diff
@@ -0,0 +1,159 @@
+        <expand macro="process_inputs"/>
+        <section name="options_advanced" title="advanced options" expanded="False">
+	    <expand macro="common_advanced"/>
+	    <param argument="--no_read_trimming" type="boolean" checked="false" truevalue="--no_read_trimming" falsevalue="" label="do not trim low quality reads, just discard them" />
```
tabs

bernt-matthias replied:
done

abretaud commented on tools/stacks2/stacks_sstacks.xml:
```diff
@@ -0,0 +1,204 @@
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/>
```
tabs

bernt-matthias replied:
done

abretaud commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,454 @@
+            <option value="--index_index">@TYPE@ is provded in FASTQ header (Illumina i5 and i7 reads) (--index_index)</option>
+            <option value="--inline_index">@TYPE@ is inline with sequence on single-end read and occurs in FASTQ header (from either i5 or i7 read) (--inline_index)</option>
+            <option value="--index_inline">@TYPE@ occurs in FASTQ header (Illumina i5 or i7 read) and is inline with sequence on single-end read (if single read data) or paired-end read (if paired data) (--index_inline)</option>
+        </expand>    
```
There are a few trailing slashes in this file, and probably others

bernt-matthias replied:
@abretaud  What do you mean with training slashes?

abretaud replied:
At the end of line 452 there are a few spaces after "</expand>", and there a few other places where this happens

bernt-matthias replied:
fixed here and everywhere else

abretaud commented on tools/stacks2/stacks_shortreads.xml:
```diff
@@ -0,0 +1,159 @@
+
+**What it does**
+
+erforms the same task as process_radtags for fast cleaning of randomly sheared genomic or transcriptomic data, not for RAD data.
```
Missing 'P'

bernt-matthias replied:
Added P

abretaud commented on tools/stacks2/stacks_sstacks.xml:
```diff
@@ -0,0 +1,204 @@
+	<expand macro="input_tags_macro"/>
+
+        <param name="popmap" type="data" format="tabular,txt" label="Population map" help="If set, matching will be done only for samples listed in this file" optional="true" argument="-M" />
+        <param argument="-x" type="boolean" checked="false" truevalue="-x" falsevalue="" label="don't verify haplotype of matching locus" />
```
double negation, I'm confused!

abretaud replied:
(By the way capitalizing all labels would be good I think)

bernt-matthias replied:
So you mean that its confusing that choosing **yes** for this option means that haplotypes are **not** verified

abretaud replied:
> So you mean that its confusing that choosing **yes** for this option means that haplotypes are **not** verified

Yep, that's it!

bernt-matthias replied:
Done (commit is coming soon .. just preparing the 2.3b upgrade and some macro simplifications).

abretaud commented on tools/stacks2/stacks_sstacks.xml:
```diff
@@ -0,0 +1,204 @@
+
+        <param name="popmap" type="data" format="tabular,txt" label="Population map" help="If set, matching will be done only for samples listed in this file" optional="true" argument="-M" />
+        <param argument="-x" type="boolean" checked="false" truevalue="-x" falsevalue="" label="don't verify haplotype of matching locus" />
+        <expand macro="gap_options_onoff"/>
```
Having a select box here looks a bit strange as there are no other options depending on it. Maybe it's worth having a standard boolean field?

bernt-matthias replied:
Not entirely sure what you mean here / which line

bernt-matthias replied:
Problem is that by IUC standards the conditionals are required to be controlled by a select. I also would prefer a boolean. We should stick with the standard OR fix the "problem" (never got the details) that led to this standard.

abretaud replied:
I meant that maybe we could replace the whole conditional by a single bool param in sstacks, and keep a conditional in cstacks and ustacks as they make use of the <yield />. But it's probably not worth the effort, I'm fine with leaving it as is

bernt-matthias replied:
then lets leave it as is :)

yvanlebras commented:
Yes, real-life can be my job ;) 


De: "Anthony Bretaudeau" <notifications@github.com> 
À: "galaxyproject/tools-iuc" <tools-iuc@noreply.github.com> 
Cc: "Yvan LE BRAS" <yvan.le-bras@mnhn.fr>, "Mention" <mention@noreply.github.com> 
Envoyé: Vendredi 18 Janvier 2019 15:40:54 
Objet: Re: [galaxyproject/tools-iuc] Stacks update (#2085) 



@abretaud commented on this pull request. 

Good work [ https://github.com/bernt-matthias | @bernt-matthias ] ! 
A few (mostly cosmetic) comments 
I didn't tried to run the tools as it's been a loooong time since I've played with radseq data, but the tools look god. Maybe [ https://github.com/yvanlebras | @yvanlebras ] will be more helpful for real-life usage? 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r248247008 | tools/stacks2/.shed.yml ] : 
> @@ -0,0 +1,19 @@
+categories:
+- Sequence Analysis
+description: Stacks is a software pipeline for building loci from short-read sequences, such as RAD-seq
+long_description: |
+  Stacks is a software pipeline for building loci from short-read sequences, such as those generated on the Illumina platform.
+  Stacks was developed to work with restriction enzyme-based data, such as RAD-seq, for the purpose of building genetic maps and conducting population genomics and phylogeography.
+
+  http://catchenlab.life.illinois.edu/stacks/
+homepage_url: http://catchenlab.life.illinois.edu/stacks/
+name: stacks 


This is not needed for suites 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r248247295 | tools/stacks2/macros.xml ] : 
> +        <option value="xhoI">xhoI</option>
+    </xml>
+
+    <!-- log file handling -->
+    <token name="@TEE_APPEND_LOG@"><![CDATA[
+        #if $output_log
+            2>> $output_log &&
+        #end if
+    ]]></token>
+    <token name="@CAT_LOG_TO_STDERR@"><![CDATA[
+        #if $output_log
+            cat $output_log 2>&1 
+        #end if
+    ]]></token>
+    <xml name="in_log">
+        <param name="add_log" type="boolean" checked="false" truevalue="yes" falsevalue="no" label="Add log output as data set" /> 


data set -> dataset 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249054645 | tools/stacks2/macros_process.xml ] : 
> +    #if str($options_advanced.adapter_2) != "":
+        --adapter_2 $options_advanced.adapter_2
+    #end if
+    #if str($options_advanced.adapter_mm) != "":
+        --adapter_mm $options_advanced.adapter_mm
+    #end if
+    ]]></token>
+
+    <!-- barcode rescue options -->
+    <xml name="rescue_barcode">
+        <conditional name="rescue_cond">
+            <param name="rescue" type="select" argument="-r" label="Rescue mutated barcodes and RAD-Tags?">
+                <option value="-r">yes</option>
+                <option value="" selected="true">no</option>
+            </param>
+	    <when value="-r"> 


tabs here and in the following line 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056078 | tools/stacks2/macros.xml ] : 
> +`Stacks manual <http://catchenlab.life.illinois.edu/stacks/manual/>`_
+
+`Stacks google group <http://groups.google.com/group/stacks-users>`_
+]]></token>
+
+    <xml name="enzymes">
+
+        <option value="">Unspecified</option>
+        <option value="aciI">aciI</option>
+        <option value="ageI">ageI</option>
+        <option value="aluI">aluI</option>
+        <option value="apaLI">apaLI</option>
+        <option value="apeKI">apeKI</option>
+        <option value="apoI">apoI</option>
+        <option value="aseI">aseI</option>
+	<option value="bamHI">bamHI</option> 


tabs (and below) 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056225 | tools/stacks2/stacks_cstacks.xml ] : 
> +#if $k_len:
+    --k_len $k_len
+#end if
+$report_mmatches
+@TEE_APPEND_LOG@ 
+@CAT_LOG_TO_STDERR@ 
+
+#if $popmap
+    ## When using a popmap, cstacks writes to the input dir
+    && mv stacks_inputs/catalog.*.tsv stacks_outputs/
+#end if
+
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/> 


tabs 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056372 | tools/stacks2/stacks_gstacks.xml ] : 
> +                    <param name="paired_select" value=""/>
+                    <!--<param name="rm_unpaired_reads" value="\-\-rm-unpaired-reads" /> removes to much of the test data and gstacks fails-->
+                    <param name="rm_pcr_duplicates" value="--rm-pcr-duplicates" />
+                </conditional>
+            </conditional>
+            <param name="add_log" value="yes" />
+            <assert_command>
+                <has_text text="-I bam_inputs" />
+                <not_has_text text="-B " />
+                <has_text text="--rm-unpaired-reads" />
+                <has_text text="--rm-pcr-duplicates" />
+            </assert_command>
+        </test>
+        <!-- refbased w popmap (here bam names need to be equal to sample names in popmap), \-\-unpaired, advanced, snp model -->
+        <test>
+		<param name="input_bam" ftype="bam" value="refmap/PopA_01.bam,refmap/PopA_02.bam"/> 


tabs 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056606 | tools/stacks2/stacks_kmerfilter.xml ] : 
> +    
+    #if $capture:
+    #if $input_type.input_type_select == "single"
+        && mv outputs/forward.discards.fq '$discarded'
+    #else
+        && mv outputs/forward.discards.fq '$discarded_pair.forward'
+        && mv outputs/reverse.discards.fq '$discarded_pair.reverse'
+    #end if
+    #end if
+
+#end if
+
+    ]]></command>
+    <inputs>
+        <expand macro="fastq_input_pair_bc"/>
+	<param name="capture" type="boolean" checked="false" truevalue="-D" falsevalue="" argument="-D" label="Capture discarded reads to a file" /> 


tabs 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056765 | tools/stacks2/stacks_procrad.xml ] : 
> +                    <expand macro="enzymes"/>
+                </param>
+            </when>
+            <when value="2">
+                <param name="enzyme" type="select" label="Enzyme" argument="-e" help="provide the restriction enzyme used" >
+                    <expand macro="enzymes"/>
+                </param>
+                <param name="enzyme2" type="select" label="Second enzyme" argument="--renz_2" help="provide the second restriction enzyme used" >
+                    <expand macro="enzymes"/>
+                </param>
+            </when>
+        </conditional>
+
+        <section name="options_advanced" title="advanced options" expanded="False">
+	    <expand macro="common_advanced"/>
+            <param argument="--bestrad" type="boolean" checked="false" truevalue="--bestrad" falsevalue="" label="library was generated using BestRAD, check for restriction enzyme on either read and potentially tranpose reads" /> 


tabs 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056832 | tools/stacks2/stacks_shortreads.xml ] : 
> +## advanced options not shared between shortreads and radtags
+$options_advanced.no_read_trimming
+$options_advanced.mate_pair
+$options_advanced.no_overhang
+
+#if $output_log
+    && mv stacks_outputs/process_shortreads.log $output_log
+#end if
+@PROCESS_FASTQ_POSTPROC@
+    ]]></command>
+
+    <inputs>
+        <expand macro="process_inputs"/>
+        <section name="options_advanced" title="advanced options" expanded="False">
+	    <expand macro="common_advanced"/>
+	    <param argument="--no_read_trimming" type="boolean" checked="false" truevalue="--no_read_trimming" falsevalue="" label="do not trim low quality reads, just discard them" /> 


tabs 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249056977 | tools/stacks2/stacks_sstacks.xml ] : 
> +#end if
+
+$x
+
+@GAP_OPTIONS_ONOFF@
+@TEE_APPEND_LOG@ 
+@CAT_LOG_TO_STDERR@
+
+#if $popmap
+    ## When using a popmap, stacks write to the input dir
+    && mv stacks_inputs/*matches.tsv stacks_outputs/
+#end if
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/> 


tabs 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249058651 | tools/stacks2/macros.xml ] : 
> +    </xml>
+
+    <xml name="barcode_encoding_single" token_type="">
+        <option value="--inline_null" selected="True">@TYPE@ is inline with sequence, occurs only on single-end read (--inline_null)</option>
+        <option value="--index_null">@TYPE@ is provded in FASTQ header (Illumina i5 or i7 read) (--index_null)</option>
+        <yield/>
+    </xml>
+
+    <xml name="barcode_encoding_pair" token_type="">
+        <expand macro="barcode_encoding_single" type="@TYPE@">    
+            <option value="--null_index">@TYPE@ is provded in FASTQ header (Illumina i7 read if both i5 and i7 read are provided) (--null_index)</option>
+            <option value="--inline_inline">@TYPE@ is inline with sequence, occurs on single and paired-end read (--inline_inline)</option>
+            <option value="--index_index">@TYPE@ is provded in FASTQ header (Illumina i5 and i7 reads) (--index_index)</option>
+            <option value="--inline_index">@TYPE@ is inline with sequence on single-end read and occurs in FASTQ header (from either i5 or i7 read) (--inline_index)</option>
+            <option value="--index_inline">@TYPE@ occurs in FASTQ header (Illumina i5 or i7 read) and is inline with sequence on single-end read (if single read data) or paired-end read (if paired data) (--index_inline)</option>
+        </expand> 


There are a few trailing slashes in this file, and probably others 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249061043 | tools/stacks2/stacks_shortreads.xml ] : 
> +            </output_collection>
+            <output_collection name="discarded_paired" type="list:paired" count="1">
+                <element name="reads">
+                    <element name="forward" file="shortreads/reads.forward.fq" ftype="fastqsanger"/>
+                    <element name="reverse" file="shortreads/reads.forward.fq" ftype="fastqsanger"/>
+                </element>
+            </output_collection>
+        </test>
+    </tests>
+    <help>
+<![CDATA[
+.. class:: infomark
+
+**What it does**
+
+erforms the same task as process_radtags for fast cleaning of randomly sheared genomic or transcriptomic data, not for RAD data. 


Missing 'P' 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249062014 | tools/stacks2/stacks_sstacks.xml ] : 
> +@GAP_OPTIONS_ONOFF@
+@TEE_APPEND_LOG@ 
+@CAT_LOG_TO_STDERR@
+
+#if $popmap
+    ## When using a popmap, stacks write to the input dir
+    && mv stacks_inputs/*matches.tsv stacks_outputs/
+#end if
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/>
+	<expand macro="input_tags_macro"/>
+
+        <param name="popmap" type="data" format="tabular,txt" label="Population map" help="If set, matching will be done only for samples listed in this file" optional="true" argument="-M" />
+        <param argument="-x" type="boolean" checked="false" truevalue="-x" falsevalue="" label="don't verify haplotype of matching locus" /> 


double negation, I'm confused! 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249063226 | tools/stacks2/stacks_sstacks.xml ] : 
> +@TEE_APPEND_LOG@ 
+@CAT_LOG_TO_STDERR@
+
+#if $popmap
+    ## When using a popmap, stacks write to the input dir
+    && mv stacks_inputs/*matches.tsv stacks_outputs/
+#end if
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/>
+	<expand macro="input_tags_macro"/>
+
+        <param name="popmap" type="data" format="tabular,txt" label="Population map" help="If set, matching will be done only for samples listed in this file" optional="true" argument="-M" />
+        <param argument="-x" type="boolean" checked="false" truevalue="-x" falsevalue="" label="don't verify haplotype of matching locus" />
+        <expand macro="gap_options_onoff"/> 


Having a select box here looks a bit strange as there are no other options depending on it. Maybe it's worth having a standard boolean field? 


In [ https://github.com/galaxyproject/tools-iuc/pull/2085#discussion_r249063601 | tools/stacks2/stacks_sstacks.xml ] : 
> +@GAP_OPTIONS_ONOFF@
+@TEE_APPEND_LOG@ 
+@CAT_LOG_TO_STDERR@
+
+#if $popmap
+    ## When using a popmap, stacks write to the input dir
+    && mv stacks_inputs/*matches.tsv stacks_outputs/
+#end if
+    ]]></command>
+
+    <inputs>
+	<expand macro="input_cat_macro"/>
+	<expand macro="input_tags_macro"/>
+
+        <param name="popmap" type="data" format="tabular,txt" label="Population map" help="If set, matching will be done only for samples listed in this file" optional="true" argument="-M" />
+        <param argument="-x" type="boolean" checked="false" truevalue="-x" falsevalue="" label="don't verify haplotype of matching locus" /> 


(By the way capitalizing all labels would be good I think) 

— 
You are receiving this because you were mentioned. 
Reply to this email directly, [ https://github.com/galaxyproject/tools-iuc/pull/2085#pullrequestreview-193091507 | view it on GitHub ] , or [ https://github.com/notifications/unsubscribe-auth/AHi1FycQU5lLiz7ofSGElRdZ0CJOpExXks5vEdz2gaJpZM4WpoDi | mute the thread ] . 

-- 
-- 
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
Yvan Le Bras, PhD @Yvan2935 <°))))>< 
Chef de projet Pole National de Données de Biodiversité 
Bureau 34, Station marine de Concarneau BP 225, 29182 Concarneau CEDEX --- MNHN PatriNat UMS 2006 Paris 
tél.: +33 (0) 2 98 50 99 35 / +33 (0) 6.10.43.96.51 
yvan.le-bras@mnhn.fr

bernt-matthias commented:
Hey @yvanlebras and @abretaud . Thanks for the comments. Will take care in a few days. Will be offline next week.

bernt-matthias commented:
I have a few changes in my local git which I did not push yet .. but its only for kmerfilter and some macro simplifications .. I think you can proceed with testing if you find the time.

bernt-matthias commented on tools/stacks2/todo.txt:
```diff
@@ -0,0 +1,44 @@
+For new tools I tried to add sections analogous to the strcuture
+of stacks help output. Since we break compatibility to the stacks 1.48
```
Will remove the todo.txt as soon as possible. I added it to the PR decription in order to not forget it.

bernt-matthias commented on tools/stacks2/.shed.yml:
```diff
@@ -0,0 +1,19 @@
+
+  http://catchenlab.life.illinois.edu/stacks/
+homepage_url: http://catchenlab.life.illinois.edu/stacks/
+name: stacks
```
`name` has been removed

bernt-matthias commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,454 @@
+        #end if
+    ]]></token>
+    <xml name="in_log">
+        <param name="add_log" type="boolean" checked="false" truevalue="yes" falsevalue="no" label="Add log output as data set" />
```
I have made this replacement everywhere

bernt-matthias commented on tools/stacks2/stacks_clonefilter.xml:
```diff
@@ -0,0 +1,165 @@
+                <option value="pair">Pair</option>
+            </param>
+            <when value="single">
+                <param name="fname" type="data" format="fastqsanger,fastqsanger.gz" label="FASTQ" />
```
solved

bernt-matthias commented on tools/stacks2/stacks_populations.xml:
```diff
@@ -0,0 +1,710 @@
+
+
+
+         <!-- TODO test wo popmap  -->
```
done

bernt-matthias commented:
Despite the last commits were quite large no functionality was changed (only 2/3 parameters that were added in 2.3).. it was mostly test and macro simplifications and improvements. 

Note: I expect a few tests still to fail here because of https://github.com/galaxyproject/planemo/issues/901 which is about to be solved here https://github.com/galaxyproject/galaxy/pull/7372 (I plan to backport this to the galaxy release that is currently used in planemo). 

2.3c bugfix release is coming https://github.com/bioconda/bioconda-recipes/pull/13821 -- should be a trivial exchange

Apart from this I would say its finished. 

What do you think: Should we prepare a workflow for this? I never did this, but it would be nice to have a workflow test as well. Somebody has experience with this and could send me some links where I could start reading?

yvanlebras commented:
Thank you @bernt-matthias for this huge work!
As mentionned before, I can make some "real" tests on it and so try to create some workflows.
Tools from TTS are updated ? I can install these tools to try ?

bernt-matthias commented:
uploading now to TTS

bernt-matthias commented:
done. I guess suite_stacks2 should install all tools.

yvanlebras commented:
Just succeed to take time to deploy a dedicated Galaxy flavour and install suite_stack2! Try to make tests and give feedbacks ASAP

yvanlebras commented:
Beginning test through https://galaxyproject.github.io/training-material/topics/sequence-analysis/tutorials/ref-based-rad-seq/tutorial.html tutorial, I have an issue on the first process radtag step... 
```
Fatal error: Exit code 1 ()
ln: failed to create symbolic link ‘stacks_inputs/EBI SRA: SRR034310 File: ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR034/SRR034310/SRR034310.fastq’: No such file or directory
```
Due to the name of the dataset (with space and other things like that). So trying after modifying the name of the dataset, I have a new error:
```
Fatal error: Exit code 127 ()
/export/galaxy-central/database/job_working_directory/000/71/tool_script.sh: line 9: process_radtags: command not found
```
So apparently there was a problem during the automatic installation of the dependencies...

I will investiguate further tomorrow!

yvanlebras commented:
Update: Don't know why but stacks conda dependencies is ok this morning... Maybe it take longer to install it... I continue to test!

yvanlebras commented:
I am now at the reference-map test. Until now, all seems to be ok. Just for reference-map, it appears to me that we only can use this tool for population usage, no option for genetic map ? Is it normal ?

yvanlebras commented:
Finally succeed to finish this https://galaxyproject.github.io/training-material/topics/sequence-analysis/tutorials/ref-based-rad-seq/tutorial.html tutorial.

It appears to me that **_populations_** didn't output `populations.fst_summary.tsv` as `populations.phistats_summary.tsv` on th history... can you confirm this ?

yvanlebras commented on tools/stacks2/stacks_denovomap.xml:
```diff
@@ -0,0 +1,262 @@
+
+    <inputs>
+        <expand macro="fastq_input"/>
+        <param argument="--popmap" type="data" format="tabular,txt" label="Population map" />
```
It appears to me that for popmap input file for refmap as denovomap we have to specify optional="true" thus for genetic map, we can specify "nothing selected"

bernt-matthias replied:
You are right. I was mislead by the command line help which lists the popmap as mandatory argument: 

`denovo_map.pl --samples dir --popmap path -o dir [--paired] (assembly options) [-X prog:"opts" ...]`

I will make them optional in both scripts.

yvanlebras commented on tools/stacks2/stacks_refmap.xml:
```diff
@@ -0,0 +1,151 @@
+
+    <inputs>
+        <expand macro="bam_input_macro"/>
+        <param argument="--popmap" type="data" format="tabular,txt" label="Population map" />
```
It appears to me that for popmap input file for refmap as denovomap we have to specify optional="true" thus for genetic map, we can specify "nothing selected"

yvanlebras commented:
A general comment: Maybe it will be better to activate creation of log files by defaults

yvanlebras commented:
I encounter difficulties with the tsv2bam tool. Trying it on single end files in this [history](https://openstack-192-168-100-96.genouest.org/u/ylebras/h/test-stacks2-genetic-map), datasetstep 218, I have this error:
```
Logging to 'stacks_inputs/tsv2bam.log'.
Configuration for this run:
  Stacks directory: 'stacks_inputs/'
  Num. samples: 20
  Paired-end reads directory: 'stacks_inputs/'

Paired-end reads files found, e.g. 'stacks_inputs/progeny_1.2.fasta'.
Loading the catalog...
Processing sample 'progeny_1'...
Error: Unrecognized paired-end read name format, at 'TTAAT_1_0012_5149_6354_1[75176]'.
Aborted.
```

It seems the tsv2bam is searching paired-end files... Maybe an issue with the created comand line AND/OR of my reads / file names ?

bernt-matthias commented:
Thanks for the comments. Will handle them as soon as possible: 

> I am now at the reference-map test. Until now, all seems to be ok. Just for reference-map, it appears to me that we only can use this tool for population usage, no option for genetic map ? Is it normal ?

I guess you are referring to the absence of the `genotypes` program. It seems that this has not been ported (yet..?):

https://groups.google.com/forum/#!topic/stacks-users/oX-T4Xefe9w/discussion
https://groups.google.com/forum/#!topic/stacks-users/x2J1AXuklTg/discussion

bernt-matthias commented:
> It appears to me that populations didn't output populations.fst_summary.tsv as populations.phistats_summary.tsv on the history... can you confirm this ?

These files are only in the output if: 
1. populations is run separately (neither refmap or denovomap output these files .. ) 
2. `Enable SNP and haplotype-based F statistics` must be set to yes
3. `Advanced options` -> `Log components of Fst/Phi_st calculations to a dataset` must be enabled

If I remember correctly, then the files are not generated if 2 and 3 are not set (which is the default in the pipelines [denovomap and refmap]). 

The reasons are here 

https://github.com/galaxyproject/tools-iuc/blob/ae0b563c09231e46355ed18f8274e032f654669d/tools/stacks2/stacks_denovomap.xml#L57

and here

https://github.com/galaxyproject/tools-iuc/blob/ae0b563c09231e46355ed18f8274e032f654669d/tools/stacks2/macros.xml#L451

Note that, actually the rules affect 3 files: 
- populations.fst_summary.tsv 
- populations.phistats_summary.tsv 
- populations.phistats.tsv

bernt-matthias commented:
> A general comment: Maybe it will be better to activate creation of log files by defaults (so by default affect the "yes" value to Add log output as dataset)

Agreed and changed.

bernt-matthias commented:
> Beginning test through https://galaxyproject.github.io/training-material/topics/sequence-analysis/tutorials/ref-based-rad-seq/tutorial.html tutorial, I have an issue on the first process radtag step...
> 
> ```
> Fatal error: Exit code 1 ()
> ln: failed to create symbolic link ‘stacks_inputs/EBI SRA: SRR034310 File: ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR034/SRR034310/SRR034310.fastq’: No such file or directory
> ```
> 
> Due to the name of the dataset (with space and other things like that). So trying after modifying the name of the dataset, I have a new error:

This is indeed a problem that I was thinking about (but https://github.com/galaxyproject/galaxy/pull/7372 seems not to be the solution). Any characters like '/' will lead to problems. Maybe we need a kind of validator for the data set names. I'm unsure if quoting might be solution, since the quoted data sets names will also appear in the output.

I guess in most applications this won't be a problem since, the data set names are created automatically, e.g. here with procrad. But we should take care of this anyway.

yvanlebras commented:
Thank you Matthias! I understand considering the input name, and this really minor as user can change the dataset name to fix that. I'm ready to test your updates when you want, after you update the testTS repo.

Thank you so much!!

bernt-matthias commented:
Hi @yvanlebras . Unfortunatelly I was not able to understand the problem in your history here https://openstack-192-168-100-96.genouest.org/u/ylebras/h/test-stacks2-genetic-map .. I guess solving this would be great.

My problem is that with Galaxy's naming scheme its difficult to find out what the inputs of the single steps were. Can you describe single steps for reproducing the error?

yvanlebras commented:
Oh yes, you mean for the tsv2bam error on dataset 218? I will have a look

yvanlebras commented:
Hi Matthias, don't hesitate to ping me if you update test TS repo and want new tests. Thanks again for your work!

bgruening commented on tools/stacks2/stacks_tsv2bam.xml:
```diff
@@ -0,0 +1,254 @@
+        <expand macro="input_stacks_macro"/>
+        <expand macro="input_cat_macro"/>
+        <expand macro="input_matches_macro"/>
+        <!-- TODO add BAM? -->
```
is this still a ToDo?

bernt-matthias replied:
left this for future contributors

bgruening commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,592 @@
+    <!-- log file handling -->
+    <token name="@TEE_APPEND_LOG@"><![CDATA[
+        #if $output_log
+            2>> $output_log &&
```
quotes?

bernt-matthias replied:
done

bgruening commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,592 @@
+    ]]></token>
+    <token name="@CAT_LOG_TO_STDERR@"><![CDATA[
+        #if $output_log
+            cat $output_log 2>&1
```
quotes?

bernt-matthias replied:
done

bgruening commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,592 @@
+        #while $identifier.endswith(('.1', '.2', '.fa', '.fq', '.fasta', '.fastq', '.gz', '.gzip', '.sam', '.bam'))
+            #set $identifier = splitext($identifier)[0]
+        #end while
+$identifier#slurp
```
indentation

bernt-matthias replied:
done

bernt-matthias replied:
have to correct my self. Can't add the spaces here, otherwise these spaces would be included in the linkname. which is created here.

bgruening commented on tools/stacks2/macros.xml:
```diff
@@ -0,0 +1,592 @@
+    ]]></token>
+    <xml name="gap_options">
+        <expand macro="gap_options_onoff">
+            <param argument="--max_gaps" type="float" value="2.0" label="Number of gaps allowed between stacks before merging"/>
```
how can gaps be a float?

bernt-matthias replied:
No idea. I checked in the stacks sources where it is definitely parsed as a double. Also the default is explicitly set to 2.0 and not just 2. 

Would be easy to make it an integer in the wrapper. But I tried to stay as close at the original as possible.

bgruening replied:
Seems reasonable!  Thanks a lot Matthias!

yvanlebras commented:
Thank you Matthias! I just finish updating all tools from testTS and begin to retest it! Here are some comments, maybe not relevant, maybe not so important, on the fly to have a basis to exchange:
- for denovomap, it seems we can't give Population filtering options:
  - -p,--min-populations — minimum number of populations a locus must be present in to process a locus (for populations; default: 1)
  - -r,--min-samples-per-pop — minimum percentage of individuals in a population required to process a locus for that population (for populations; default: 0)
  - we can also propose a way to use X parameter so the user can give option for each part of the denovomap pipeline. THIS IS NOT MANDATORY, maybe just a big + . X — additional options for specific pipeline components, e.g. -X "populations: --min-maf 0.05".
- for tsv2bam, it seems to me (sorry don't take time to look at the command line) that the -R is given by default isn't it ? Because Stacks is searching for pair-end data even if I only have single end? It appears to me that using the `Short read data from individuals` is mis-related to -R option isn't it ?

bernt-matthias commented:
> Thank you Matthias! I just finish updating all tools from testTS and begin to retest it! Here are some comments, maybe not relevant, maybe not so important, on the fly to have a basis to exchange:

Thanks for the comments. They are definitely valuable.

>     * for denovomap, it seems we can't give Population filtering options:
>       
>       * -p,--min-populations — minimum number of populations a locus must be present in to process a locus (for populations; default: 1)
>       * -r,--min-samples-per-pop — minimum percentage of individuals in a population required to process a locus for that population (for populations; default: 0)

Don't know how there slipped through.... I just added them.

Meanwhile I also removed the `--paired` option from the options accessible by the user, its now just set if the user uploads paired data.

>       * we can also propose a way to use X parameter so the user can give option for each part of the denovomap pipeline. THIS IS NOT MANDATORY, maybe just a big + . X — additional options for specific pipeline components, e.g. -X "populations: --min-maf 0.05".

I also thought about this earlier. I'm against it for the following reason: 

- users that want those options may easily construct a workflow consisting of the separate steps (we might help them by extending the documentations sketching the pipelines .. as I did for dada2 .. or just provide the workflow via the toolshed). Actually I thought about dropping the Galaxy support for the two 
perl scripts at all :) 

- if someone still wants to provide the -X option than this should be done by having the in/outputs of every tool in macros that can then be used in the actual tools and the pipeline wrappers. 

>     * for tsv2bam, it seems to me (sorry don't take time to look at the command line) that the -R is given by default isn't it ? Because Stacks is searching for pair-end data even if I only have single end? It appears to me that using the `Short read data from individuals` is mis-related to `--pe-reads-dir` option isn't it ?

The generated command lines for the three tests are: 

1. `tsv2bam -P stacks_inputs -s 'PopA_01' -s 'PopA_02' -t ${GALAXY_SLOTS:-1}`
2. `tsv2bam -P stacks_inputs -M '/tmp/tmpf6lcxe0l/files/e/b/5/dataset_eb5f0a84-ea84-4f38-952f-864b1f29a6e0.dat' --pe-reads-dir stacks_inputs -t ${GALAXY_SLOTS:-1}`
3. `tsv2bam -P stacks_inputs -s 'PopA_01' -s 'PopA_02' --pe-reads-dir stacks_inputs -t ${GALAXY_SLOTS:-1}`

`--pe-reads-dir`  (ie. `-R`) is only given in the case that paired end data is given (note that `-P` and `--pe-reads-dir` point to the same dir -- which seems to be OK).

I think that is correct, isn't it?

bernt-matthias commented:
> It appears to me that using the `Short read data from individuals` is mis-related to `--pe-reads-dir` option isn't it ?
> 
>     * looking at the Galaxy scripts, it seems an error can be there, maybe into `#for $sample in $input_type.fqinputs:` ? :

I don't quite get this comment.

yvanlebras commented:
Thank you Matthias for your answer!
- Thank you for ading population related options on denovomap and PE improvements !
- Totally agree with you for the -X things
- Concerning `tsv2bam` I only have 2 possibilities (as seen in the above screenshot)
![image](https://user-images.githubusercontent.com/7910679/60080316-f0293400-972f-11e9-92f4-9bd1842755f5.png)
 and  selecting the only compatible "single end" option, `single end or forward reads`, I have an error with this stdout:
```
Logging to 'stacks_inputs/tsv2bam.log'.
Configuration for this run:
  Stacks directory: 'stacks_inputs/'
  Population map: '/export/galaxy-central/database/files/002/dataset_2817.dat'
  Num. samples: 22
  Paired-end reads directory: 'stacks_inputs/'

Paired-end reads files found, e.g. 'stacks_inputs/female.2.fasta'.
Loading the catalog...
Processing sample 'female'...
Error: Unrecognized paired-end read name format, at 'GGAAG_2_0098_639_687_1[75176]'.
Aborted.
```
This can also be due to input filenames maybe, will check that

bernt-matthias commented:
OK. Got it. The wording war wrong here. I changed the first option to `reverse reads` (so the idea is that you may provide the reverse reads as list or multiple selection). 

I also added to the help text: `Leave selection empty if you analyse single end data.`

Is the data that you provided the reverse reads or did you try to provide forward/single end reads?

yvanlebras commented:
Thank you so much! I will try to test ASAP!

yvanlebras commented:
Don't hesitate to ping me when you update TTS repos

bernt-matthias commented:
all updated @yvanlebras

yvanlebras commented:
Thank you very very much!!

yvanlebras commented:
OK for denovo (now with population options) !
OK for tsv2bam (now from single end data) !
Thank you.
I will try to test others stacks part and come back ASAP

yvanlebras commented:
Ok, Almost finished to verify all existing Galaxy training material workflows using the new version of Stacks. It seems that some files are missing, using [Stacks2: populations](https://testtoolshed.g2.bx.psu.edu/view/matthias/stacks2_populations), and specifying F based statistics options, on the log there is mention of 2 .tsv output files who are not on the output history datasets:

```
Population pair divergence statistics (more in populations.fst_summary.tsv and populations.phistats_summary.tsv
```

bernt-matthias commented:
Jep, these output data sets or only added if 

1. `Enable SNP and haplotype-based F statistics` is set to yes
2. The option `Advanced options -- Log components of Fst/Phi_st calculations to a dataset` (--log_fst_comp) is checked. 

We could enable the second by default if you like (I have no preference).

yvanlebras commented:
Right, sorry! So all is ok for me! THANK YOU so much!!!!! You rocks Matthias!

bernt-matthias commented:
Cool. Thanks for you efforts.

yvanlebras commented:
Really nothing compared to yours! You are Terrific! Next step dada2 ;) ?

bernt-matthias commented:
Sounds great to me.

abretaud commented:
Awesome work, thanks a lot!

yvanlebras commented:
![Uploading IMG_20190505_091301.jpg…](A lotte ?) ;)

yvanlebras commented:
![A lotte ? ;)](https://user-images.githubusercontent.com/7910679/60437177-0e9a9e00-9c0e-11e9-8a0b-1902f6021f02.jpg)

abretaud commented:
A Freiburg lotte? scarry

bgruening commented:
I got this right you need this on Thursday? I will try to install those tool over the night, is that ok?

@yvanlebras in the RAD-seq category?

yvanlebras commented:
It will be super genial mega cool bro! Yep, RAD(seq category is perfect! Thanks!
